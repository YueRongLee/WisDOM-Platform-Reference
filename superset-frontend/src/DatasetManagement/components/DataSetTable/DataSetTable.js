/* eslint-disable react-hooks/exhaustive-deps */
/* eslint-disable no-console */
/* eslint-disable no-restricted-imports */
import React, { useEffect, useState } from 'react';
// import PropTypes from 'prop-types';
import {
  ExclamationCircleOutlined,
  FileSearchOutlined,
  FileDoneOutlined,
} from '@ant-design/icons';
import { Input, Select, Modal, Button, Tooltip } from 'antd';
import moment from 'moment';
import TableModal from '../TableModal/TableModal';
import { ROLE_TYPE, ROLEPERMISSION } from '~~constants/index';
import { TableApi, UserApi } from '~~apis/';
import ApprovalModal from '../ApprovalModal/ApprovalModal';
import { useModal } from '~~hooks/';

import * as Style from './style';

const { Option } = Select;

const { Search } = Input;

const { confirm } = Modal;

const defaultPagination = {
  current: 1,
  pageSize: 10,
  total: 0,
};

const defaultSorter = {
  field: 'tableName',
  order: 'descend',
};

const DataSetTable = () => {
  const [tableList, setTableList] = useState([]);
  const [changeList, setChangeList] = useState([]);
  const [memberList, setMemberList] = useState([]);
  const [loading, setLoading] = useState(false);
  const [keyword, setKeyword] = useState('');
  const [pagination, setPagination] = useState(defaultPagination);
  const [sorter, setSorter] = useState(defaultSorter);

  const tableModal = useModal();
  const approvalModal = useModal();
  const SYSTEMLIST = JSON.parse(localStorage.getItem('rolePermission'));
  const onChangeInput = e => {
    setKeyword(e.target.value);
  };

  const columns = (list, handleChange, sorter, tableModal, approvalModal) => [
    {
      title: 'Table Name',
      dataIndex: 'tableName',
      width: '15%',
      sortOrder: sorter.field === 'tableName' ? sorter.order : false,
      sorter: true,
    },
    {
      title: 'Source',
      dataIndex: 'source',
      width: '10%',
      sortOrder: sorter.field === 'source' ? sorter.order : false,
      sorter: true,
    },
    {
      title: 'Consume Time',
      dataIndex: 'consumeTime',
      width: '10%',
      sortOrder: sorter.field === 'consumeTime' ? sorter.order : false,
      sorter: true,
      render: value => moment(value).format('YYYY/MM/DD HH:MM'),
    },
    {
      title: 'Update Time',
      dataIndex: 'updateTime',
      width: '10%',
      sortOrder: sorter.field === 'updateTime' ? sorter.order : false,
      sorter: true,
      render: value => moment(value).format('YYYY/MM/DD HH:MM'),
    },
    {
      title: 'Data Domain',
      dataIndex: 'category',
      width: '8%',
      sortOrder: sorter.field === 'category' ? sorter.order : false,
      sorter: true,
    },
    {
      title: 'Owner',
      dataIndex: 'owner',
      width: '16%',
      sortOrder: sorter.field === 'owner' ? sorter.order : false,
      sorter: true,
      render: (value, record) => (
        <Select
          disabled={
            !localStorage.getItem('role').includes(ROLE_TYPE.DATA_MASTER) ||
            !ROLEPERMISSION.checkPermission(
              SYSTEMLIST,
              ROLEPERMISSION.setting.datasetMangement.edit.value,
            )
          }
          showSearch
          value={value}
          style={{ width: 200 }}
          onChange={v => handleChange(v, record)}
          filterOption={(input, option) =>
            option.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
          }
        >
          {list.map(item => (
            <Option value={item.userId}>{item.userNameEn}</Option>
          ))}
        </Select>
      ),
    },
    {
      title: 'Total Access',
      dataIndex: 'totalAccess',
      width: '6%',
      sortOrder: sorter.field === 'totalAccess' ? sorter.order : false,
      sorter: true,
    },
    {
      title: 'Application Record',
      dataIndex: 'record',
      width: '10%',
      render: (value, record) => (
        <div
          style={{
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
          }}
        >
          {record.admin ? (
            <Tooltip placement="top" title="Application Record">
              <FileSearchOutlined
                style={{ fontSize: 20 }}
                onClick={() => tableModal.openModal(record)}
              />
            </Tooltip>
          ) : null}
        </div>
      ),
    },
    {
      title: ' Approval Record',
      dataIndex: 'record',
      width: '10%',
      render: (value, record) => (
        <div
          style={{
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
          }}
        >
          {record.dataOwner ? (
            <Tooltip placement="top" title="Application Record">
              <FileDoneOutlined
                style={{ fontSize: 20 }}
                onClick={() => approvalModal.openModal(record)}
              />
            </Tooltip>
          ) : null}
        </div>
      ),
    },
  ];

  const handleChangeSelect = (value, record) => {
    const tempTableList = [...tableList].map(item => {
      if (item.tableName === record.tableName) {
        return {
          ...item,
          owner: value,
        };
      }
      return item;
    });
    setTableList(tempTableList);
    if (changeList.length > 0) {
      const filterList = changeList.filter(
        item => item.tableName !== record.tableName,
      );
      const temp = {
        changeOwner: value,
        tableName: record.tableName,
      };
      setChangeList([...filterList, temp]);
    } else {
      const temp = {
        changeOwner: value,
        tableName: record.tableName,
      };
      setChangeList([temp]);
    }
  };

  const getUserList = async () => {
    try {
      const result = await UserApi.getUserList();
      setMemberList(result);
    } catch (e) {
      console.log(e);
    }
  };

  const getDataSetList = async (page, tempSorter) => {
    try {
      setLoading(true);
      const payload = {
        searchKey: keyword,
        page: page || pagination.current,
        pageSize: pagination.pageSize,
        sorter: tempSorter || sorter,
      };
      const result = await TableApi.getTableInfo(payload);
      setTableList(result.tableOwnerList);
      setPagination({
        ...pagination,
        total: result.pageInfo.total,
        current: page,
      });
    } catch (e) {
      console.log(e);
    } finally {
      setChangeList([]);
      setLoading(false);
    }
  };

  const onSearch = () => {
    getDataSetList(1);
  };

  // const onChangePage = page => {
  //   setPagination({
  //     ...pagination,
  //     current: page,
  //   });
  //   getDataSetList(page);
  // };

  const saveData = async () => {
    try {
      setLoading(true);
      const data = {
        tableWithOwnerChangeList: changeList,
      };
      const result = await TableApi.changeOwner(data);
      setTableList(result.tableOwnerList);
      setPagination({
        ...pagination,
        total: result.pageInfo.total,
        current: 1,
      });
      setKeyword('');
      getDataSetList();
    } catch (e) {
      console.log(e);
    } finally {
      setChangeList([]);
      setLoading(false);
    }
  };

  const showConfirm = () => {
    confirm({
      title: 'Do you Want to save these change items?',
      icon: <ExclamationCircleOutlined />,
      // content: 'Some descriptions',
      onOk() {
        saveData();
      },
      onCancel() {
        getDataSetList();
      },
    });
  };

  useEffect(() => {
    getUserList();
    getDataSetList();
  }, []);

  const onChange = (pagination, filters, tempSorter) => {
    getDataSetList(pagination.current, {
      field: tempSorter.field,
      order: tempSorter.order || 'ascend',
    });
    setSorter({ field: tempSorter.field, order: tempSorter.order || 'ascend' });
  };

  return (
    <Style.Container>
      <div
        className="DataSetTable"
        style={{
          display: 'flex',
          justifyContent: 'flex-end',
          marginBottom: 10,
        }}
      >
        <Search
          id="test-keyword"
          placeholder="input search text"
          onSearch={onSearch}
          value={keyword}
          style={{
            marginRight: 10,
            width: 200,
          }}
          onChange={onChangeInput}
        />
        <Button
          id="test-button"
          style={{
            width: 200,
          }}
          onClick={showConfirm}
          type="primary"
          disabled={changeList.length === 0}
        >
          Save Data
        </Button>
      </div>
      <Style.DataSetTable
        columns={columns(
          memberList,
          handleChangeSelect,
          sorter,
          tableModal,
          approvalModal,
        )}
        dataSource={loading ? [] : tableList}
        pagination={{
          current: pagination.current,
          total: pagination.total,
          pageSize: 10,
          // onChange: onChangePage,
        }}
        onChange={onChange}
        showSizeChanger={false}
        scroll={{ y: '55vh' }}
        rowKey="guid"
        loading={loading}
      />
      <TableModal modal={tableModal} />
      <ApprovalModal modal={approvalModal} />
    </Style.Container>
  );
};

DataSetTable.propTypes = {};

DataSetTable.defaultProps = {};

export default DataSetTable;
