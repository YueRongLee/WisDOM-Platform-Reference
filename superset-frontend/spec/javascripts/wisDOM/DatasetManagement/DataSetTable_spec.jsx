import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
// import renderer from 'react-test-renderer';
import DataSetTable from 'src/DatasetManagement/components/DataSetTable/DataSetTable';
import { act } from 'react-dom/test-utils';
// import { setHookState } from '../../../helpers/hook';

Enzyme.configure({ adapter: new Adapter() });

describe('DataSetTable', () => {
  it('is valid', () => {
    const wrapper = mount(<DataSetTable />); // mount/render/shallow when applicable

    expect(wrapper.find('.DataSetTable')).toExist();
  });
});

describe('Test TypeInInput', () => {
  const showConfirm = jest.fn();
  it('change event', () => {
    const onChangeInput = jest.fn();
    const wrapper = shallow(<DataSetTable onChange={onChangeInput} />);
    wrapper.find('#test-keyword').simulate('change', {
      target: {
        id: 'test-keyword',
        value: 'EE',
      },
    });
    expect(wrapper.find('#test-keyword').props().value).toEqual('EE');
  });

  it('click event', () => {
    const wrapper = mount(<DataSetTable onClick={showConfirm} />);
    // wrapper.find('#test-button').simulate('click');
    wrapper
      .find('#test-button')
      .at(0)
      .simulate('click', { target: { id: '#test-button' } });
    // wrapper.update();
    // expect(showConfirm.mock.calls.length).toEqual(1);

    // wrapper.find('#test-button').props().onClick();

    // act(() => {
    //   // Invoke the button's click handler, but this time directly, instead of
    //   // via an Enzyme API
    //   onClickConfirm();
    // });
    // Refresh Enzyme's view of the output
    wrapper.update();
    expect(showConfirm).toHaveBeenCalled();
  });
});
