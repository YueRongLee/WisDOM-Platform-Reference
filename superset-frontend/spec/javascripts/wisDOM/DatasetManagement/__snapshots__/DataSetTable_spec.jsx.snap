// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSetTable should render properly 1`] = `
<div
  className="sc-bdfBwQ jkHGLF"
>
  <div
    className="DataSetTable"
    style={
      Object {
        "display": "flex",
        "justifyContent": "flex-end",
        "marginBottom": 10,
      }
    }
  >
    <div
      data-testid="keyword"
    />
    <span
      className="ant-input-group-wrapper ant-input-search"
      style={
        Object {
          "marginRight": 10,
          "width": 200,
        }
      }
    >
      <span
        className="ant-input-wrapper ant-input-group"
      >
        <input
          className="ant-input"
          id="keyword"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="input search text"
          style={null}
          type="text"
          value=""
        />
        <span
          className="ant-input-group-addon"
        >
          <button
            className="ant-btn ant-btn-icon-only ant-input-search-button"
            onClick={[Function]}
            onMouseDown={[Function]}
            type="button"
          >
            <span
              aria-label="search"
              className="anticon anticon-search"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="search"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                />
              </svg>
            </span>
          </button>
        </span>
      </span>
    </span>
    <button
      className="ant-btn ant-btn-primary"
      disabled={true}
      onClick={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
      type="button"
    >
      <span
        className="ant-btn-loading-icon"
      >
        <span
          aria-label="loading"
          className="anticon anticon-loading anticon-spin"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="loading"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
            />
          </svg>
        </span>
      </span>
      <span>
        Save Data
      </span>
    </button>
  </div>
  <div
    className="ant-table-wrapper sc-gsTCUz iCwCSC"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table ant-table-empty ant-table-fixed-header"
        >
          <div
            className="ant-table-container"
          >
            <div
              className="ant-table-header"
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <table
                style={
                  Object {
                    "tableLayout": "fixed",
                    "visibility": null,
                  }
                }
              >
                <colgroup />
                <thead
                  className="ant-table-thead"
                >
                  <tr>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      Dataset
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      rowSpan={null}
                      style={Object {}}
                    >
                      Owner
                    </th>
                  </tr>
                </thead>
              </table>
            </div>
            <div
              className="ant-table-body"
              onScroll={[Function]}
              style={
                Object {
                  "maxHeight": "60vh",
                  "overflowY": "scroll",
                }
              }
            >
              <table
                style={
                  Object {
                    "tableLayout": "fixed",
                  }
                }
              >
                <colgroup>
                  <col
                    style={
                      Object {
                        "minWidth": "80%",
                        "width": "80%",
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": "20%",
                        "width": "20%",
                      }
                    }
                  />
                </colgroup>
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    aria-hidden="true"
                    className="ant-table-measure-row"
                    style={
                      Object {
                        "fontSize": 0,
                        "height": 0,
                      }
                    }
                  >
                    <td
                      style={
                        Object {
                          "border": 0,
                          "height": 0,
                          "padding": 0,
                        }
                      }
                    >
                      <div
                        style={
                          Object {
                            "height": 0,
                            "overflow": "hidden",
                          }
                        }
                      >
                         
                      </div>
                    </td>
                    <td
                      style={
                        Object {
                          "border": 0,
                          "height": 0,
                          "padding": 0,
                        }
                      }
                    >
                      <div
                        style={
                          Object {
                            "height": 0,
                            "overflow": "hidden",
                          }
                        }
                      >
                         
                      </div>
                    </td>
                  </tr>
                  <tr
                    className="ant-table-placeholder"
                    style={
                      Object {
                        "display": null,
                      }
                    }
                  >
                    <td
                      className="ant-table-cell"
                      colSpan={2}
                      rowSpan={null}
                      style={Object {}}
                    >
                      <div
                        className="ant-empty ant-empty-normal"
                      >
                        <div
                          className="ant-empty-image"
                        >
                          <svg
                            className="ant-empty-img-simple"
                            height="41"
                            viewBox="0 0 64 41"
                            width="64"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <g
                              fill="none"
                              fillRule="evenodd"
                              transform="translate(0 1)"
                            >
                              <ellipse
                                className="ant-empty-img-simple-ellipse"
                                cx="32"
                                cy="33"
                                rx="32"
                                ry="7"
                              />
                              <g
                                className="ant-empty-img-simple-g"
                                fillRule="nonzero"
                              >
                                <path
                                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                />
                                <path
                                  className="ant-empty-img-simple-path"
                                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                />
                              </g>
                            </g>
                          </svg>
                        </div>
                        <div
                          className="ant-empty-description"
                        >
                          No Data
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
